evm/src/lib.rs

```diff
@@ -1,3 +1,153 @@
+//!
+//! Following the [zk-EVM classification of Vitalik](https://vitalik.eth.limo/general/2022/08/04/zkevm.html),
+//! the plonky2_evm crate aims at providing an efficient solution for the problem of generating cryptographic
+//! proofs of Ethereum-like transactions with *full Ethereum capability*.
+//!
+//! To this end, the plonky2 zk-EVM is tailored for an AIR-based STARK system satisfying degree 3 constraints,
+//! with support for recursive aggregation leveraging plonky2 circuits with FRI-based plonkish arithmetization.
+//! These circuits require a one-time, offline preprocessing phase.
+//! See the [`fixed_recursive_verifier`] module for more details on how this works.
+//! These preprocessed circuits are gathered within the [`AllRecursiveCircuits`] prover state,
+//! and can be generated as such:
+//!
+//! ```ignore
+//! // Specify the base field to use.
+//! type F = GoldilocksField;
+//! // Specify the extension degree to use.
+//! const D: usize = 2;
+//! // Specify the recursive configuration to use, here leveraging Poseidon hash
+//! // over the Goldilocks field both natively and in-circuit.
+//! type C = PoseidonGoldilocksConfig;
+//!
+//! let all_stark = AllStark::<F, D>::default();
+//! let config = StarkConfig::standard_fast_config();
+//!
+//! // Generate all the recursive circuits needed to generate succinct proofs for blocks.
+//! // The ranges correspond to the supported table sizes for each individual STARK component.
+//! let prover_state = AllRecursiveCircuits::<F, C, D>::new(
+//!     &all_stark,
+//!     &[16..25, 10..20, 12..25, 14..25, 9..20, 12..20, 17..30],
+//!     &config,
+//! );
+//! ```
+//!
+//! # Inputs type
+//!
+//! Transactions need to be processed into an Intermediary Representation (IR) format for the prover
+//! to be able to generate proofs of valid state transition. This involves passing the encoded transaction,
+//! the header of the block in which it was included, some information on the state prior execution
+//! of this transaction, etc.
+//! This intermediary representation is called [`GenerationInputs`].
+//!
+//!
+//! # Generating succinct proofs
+//!
+//! ## Transaction proofs
+//!
+//! To generate a proof for a transaction, given its [`GenerationInputs`] and an [`AllRecursiveCircuits`]
+//! prover state, one can simply call the [prove_root](AllRecursiveCircuits::prove_root) method.
+//!
+//! ```ignore
+//! let mut timing = TimingTree::new("prove", log::Level::Debug);
+//! let kill_signal = None; // Useful only with distributed proving to kill hanging jobs.
+//! let (proof, public_values) =
+//!     prover_state.prove_root(all_stark, config, inputs, &mut timing, kill_signal);
+//! ```
+//!
+//! This outputs a transaction proof and its associated public values. These are necessary during the
+//! aggregation levels (see below). If one were to miss the public values, they are also retrievable directly
+//! from the proof's encoded public inputs, as such:
```

Comment:
If the PVs are included in the proof, why are they so often passed around separately? Is it expensive to recover them from the proof's encoded public values?

Reply:
It's more as a matter of convenience than anything else. It shouldn't be too heavy to retrieve them from the encoded public inputs, but as they are there anyway, it cost nothing to attach them to the output proof. I guess to be a tad cleaner, we could have a custom proof type, like `EvmProof` that consists in a `ProofWithPublicInputs` and `PublicValues`, to hide the tuple behind.

Reply:
It's aesthetically a bit annoying to pass around duplicate data, but if it's convenient as-is I'll shut up.